File: chapter 06/recipe 6.8/DniNetExceptions/DniNetExceptionsObj.cs
using System;
using System.Runtime.InteropServices;

namespace DniNetExceptions
{
	public class CustomException : ApplicationException
	{
		public CustomException(string msg)
			: base(msg)
		{
		}
	}

	public class CustomExceptionWithHResult : ApplicationException
	{
		public CustomExceptionWithHResult(string msg)
			: base(msg)
		{
			this.HResult = unchecked((int)0x80040301);
		}
	}

	public interface IExceptions
	{
		void ThrowStandardException();
		void ThrowCustomException();
		void ThrowCustomExceptionWithHResult();
	}

	[ClassInterface(ClassInterfaceType.None)]
	public class DniNetExceptionsObj : IExceptions
	{
		public void ThrowStandardException()
		{
			throw new ApplicationException(
				"This is an ApplicationException");
		}

		public void ThrowCustomException()
		{
			throw new CustomException("My custom exception message");
		}

		public void ThrowCustomExceptionWithHResult()
		{
			throw new CustomExceptionWithHResult(
				"My custom exception message with HResult");
		}
	}
}

File: chapter 06/recipe 6.8/DniNetExceptionsVB/DniNetExceptionsObj.vb
Imports System.Runtime.InteropServices

Public Class CustomException
    Inherits ApplicationException

    Public Sub New(ByVal msg As String)
        MyBase.New(msg)
    End Sub
End Class

Public Class CustomExceptionWithHResult
    Inherits ApplicationException

    Public Sub New(ByVal msg As String)
        MyBase.New(msg)
        Me.HResult = &H80040301
    End Sub
End Class

Public Interface IExceptions
    Sub ThrowStandardException()
    Sub ThrowCustomException()
    Sub ThrowCustomExceptionWithHResult()
End Interface

<ClassInterface(ClassInterfaceType.None)> _
Public Class DniNetExceptionsObj
    Implements IExceptions

    Public Sub ThrowStandardException() _
            Implements IExceptions.ThrowStandardException
        Throw New ApplicationException( _
            "This is an ApplicationException")
    End Sub

    Public Sub ThrowCustomException() _
            Implements IExceptions.ThrowCustomException
        Throw New CustomException( _
            "My custom exception message")
    End Sub

    Public Sub ThrowCustomExceptionWithHResult() _
            Implements IExceptions.ThrowCustomExceptionWithHResult
        Throw New CustomExceptionWithHResult( _
            "My custom exception message with HResult")
    End Sub

End Class
